netket/graph/grid.py

```diff
@@ -122,28 +186,132 @@ def __repr__(self):
+        if len(dims) < 2:
+            raise ValueError(f"Rotations not defined for 1d systems")
+        if _np.any(axes) > len(dims) - 1:
+            raise ValueError(f"Axis specified not in dims")
+
+        if self.length[axes[0]] == self.length[axes[1]]:
+            basis = (range(0, 4, period), [axes])
+        else:
+            basis = (range(0, 4, 2 * period), [axes])
+
+        rotations = itertools.product(*basis)
+        next(rotations)
+
+        rotations = [PlanarRotation(num, ax, dims) for (num, ax) in rotations]
+
+        return SymmGroup([Identity()] + rotations, graph=self)
+
+    def axis_reflection(self, axis: int = 0) -> SymmGroup:
+        """
+        Returns SymmGroup consisting of identity and the lattice
+        reflected about the hyperplane axis = 0
+
+        Arguments:
+            axis: Axis to be reflected about
+        """
+        if abs(axis) > len(self.length) - 1:
+            raise ValueError(f"Axis specified not in dims")
+
+        dims = tuple(self.length)
+        return SymmGroup([Identity(), Reflection(axis, dims)], graph=self)
+
+    def rotations(
+        self, period: int = 1, *, remove_duplicates: bool = True
+    ) -> SymmGroup:
+        """
+        Returns all possible rotation symmetries of the lattice.
+
+        The rotations are a subset of the permutations returned by
+        `self.automorphisms()`.
+
+        Arguments:
+            period: Period of the rotations; should be a divisor of 4.
+            remove_duplicates: Only include unique rotations.
         """
-        basis = [
-            range(l) if is_per else range(1)
-            for l, is_per in zip(self.length[::-1], self.pbc[::-1])
-        ]
+        axes = itertools.combinations(range(len(self.length)), 2)
+        group = SymmGroup([Identity()], graph=self)
+
+        for axs in axes:
+            group = group @ self.planar_rotation(axs, period)
+
+        if remove_duplicates:
+            return group.remove_duplicates()
+        else:
+            return group
 
-        translation_group = itertools.product(*basis)
-        identity = _np.array(list(self.nodes())).reshape(*self.length[::-1])
+    def space_group(self, *, remove_duplicates: bool = True) -> SymmGroup:
+        """
+        Returns the full space group of the lattice.
 
-        def translate(el, sites):
-            for i, n in enumerate(el):
-                sites = _np.roll(sites, shift=n, axis=i)
-            return sites.ravel().tolist()
+        The space group is a subset of the permutations returned by
+        `self.automorphisms()`.
 
-        return [translate(el, identity) for el in translation_group]
+        Arguments:
+            remove_duplicates: Only include unique space group elements.
+        """
+        group = self.rotations() @ self.axis_reflection()
+        if remove_duplicates:
+            return group.remove_duplicates()
+        else:
+            return group
```

Comment:
@chrisrothUT Not sure if this check is necessary, combining the (deduplicated) rotations with a single reflection may be always safe.

Reply:
I believe that is the case.

I think the space group can also be thought of as permuting axes of equal length + reflections along all axes.

The duplicates arise since there are multiple ways to achieve an n axis permutation by combining permutations of two axes. All of the permutation algebra is in the rotation group algebra

Reply:
Okay, then we agree, as does this test:
https://github.com/netket/netket/blob/10cd75bbf32843dc5ed8e3283419afd805423b40/Test/Graph/test_graph.py#L312-L319